### PR TITLE
chore: Add xml parser verify-links check

### LINKED ISSUE #757: Support XML files for sitemaps for the verify-links check
### What is this task about?
Currently the verify-link will fail, when using a sitemap.xml instead of a sitemap.txt file. Since sitemap.xml files are probably even more common than sitemap.txt files this should be addressed. 

### What needs to be done to complete this task?

- [x] Figure out how the parsing of the sitemap file is currently done
- [x] Add an option to parse XML files as sitemap files
- [x] Make the option work with the command line tool

### PR DESCRIPTION
**Pull request by dotKuro**
Should fix #757

### CODE REVIEW

**Review by goloroden — CHANGES_REQUESTED** [review]
Hi @dotKuro 👋 

This looks good, however, I have a few minor comments.
Can you please have a look at them?

**Comment by dotKuro** [comment]
Thanks for the quick reply! I fixed the issues now. 👍

**Comment by dotKuro** [comment]
When I was trying to contribute the correct typing to the @types/xml2js package, as we discussed yesterday, I found that the library has a promise based function as well. I swapped it in instead of using promisify.

File: test/cli/getUrlsFromSitemapTests.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
+import { assert } from 'assertthat';
+import { isolated } from 'isolated';
+import path from 'path';
+import { writeFileSync } from 'fs';
+import { getUrlsFromSitemap, parseSitemapTXT, parseSitemapXML } from '../../lib/cli/verifyLinks/getUrlsFromSitemap';
+import { html, stripIndent } from 'common-tags';
+
+suite('parseSitemapXML', (): void => {
+  test('returns all links from the xml file.', async (): Promise<void> => {
```
Comment (goloroden) [inline test/cli/getUrlsFromSitemapTests.ts:9]: I think we should uppercase the file endings in test descriptions, so this should be `XML` (and further down below `TXT`). I will not add a comment for every single test, but I guess you will look through all of them anyway 😊

File: test/cli/getUrlsFromSitemapTests.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    const sitemapContent = {
+      urlset: {
+        url: [
+          {
+            loc: [ 'https://example.com/' ],
+            lastmod: [ '...' ]
```
Comment (goloroden) [inline test/cli/getUrlsFromSitemapTests.ts:15]: Just for information: What would be a reasonable value for this field?

File: test/cli/getUrlsFromSitemapTests.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  });
+
+  test('ignores blank lines.', async (): Promise<void> => {
+    const sitemapContent = stripIndent`
+      https://example.com/
+
+      https://example.com/examplepage1
+      https://example.com/examplepage2
+       
+      `;
```
Comment (goloroden) [inline test/cli/getUrlsFromSitemapTests.ts:54]: I originally thought that this blank line is not needed – until I read the test description. Maybe we can highlight this a little bit more (although I don't have an idea, how, TBH).

File: test/cli/getUrlsFromSitemapTests.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        https://example.com/examplepage2`;
+
+    assert.
+      that(parseSitemapTXT(sitemapContent)).
+      is.equalTo([ 'https://example.com/', 'https://example.com/examplepage1', 'https://example.com/examplepage2' ]);
+  });
+});
+
+suite('getUrlsFromSitemap', (): void => {
+  test('parses XML file correctly.', async (): Promise<void> => {
```
Comment (goloroden) [inline test/cli/getUrlsFromSitemapTests.ts:74]: Either `XML files` or `the XML file`, I'd say.

File: test/cli/getUrlsFromSitemapTests.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+          </loc>
+          <lastmod>
+            ...
+          </lastmod>
+        </url>
+    </urlset>`;
+    const tempDirectory = await isolated();
+    const tempFile = path.join(tempDirectory, 'tempFile');
+
+    writeFileSync(tempFile, validXMLData);
```
Comment (goloroden) [inline test/cli/getUrlsFromSitemapTests.ts:99]: Can we please replace these `...Sync` calls by their promise-based counterparts from `fs.promises`?

File: test/cli/getUrlsFromSitemapTests.ts [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+    const tempFile = path.join(tempDirectory, 'tempFile');
+
+    writeFileSync(tempFile, validTXTData);
+
+    const urls = await getUrlsFromSitemap({ sitemapPath: tempFile });
+
+    assert.that(urls).is.equalTo([ 'https://example.com/', 'https://example.com/examplepage' ]);
+  });
+
+  test('throws error on valid xml file with wrong structure.', async (): Promise<void> => {
```
Comment (goloroden) [inline test/cli/getUrlsFromSitemapTests.ts:120]: I'd suggest to add an `an` between `throws` and `error`.

File: test/cli/getUrlsFromSitemapTests.ts [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    writeFileSync(tempFile, invalidXMLData);
+
+    writeFileSync(tempFile, invalidXMLData);
+
+    try {
+      await getUrlsFromSitemap({ sitemapPath: tempFile });
+    } catch {
+      errorThrown = true;
+    }
+    assert.that(errorThrown).is.true();
```
Comment (goloroden) [inline test/cli/getUrlsFromSitemapTests.ts:139]:
```suggestion
    await assert.that(async (): Promise<void> => {
      await getUrlsFromSitemap({ sitemapPath: tempFile });
    }).is.throwingAsync();
```

**Review by goloroden — CHANGES_REQUESTED** [review]
Hey @dotKuro 👋 

It's me, once again… I have reviewed your changes, and I still have a few things. But, we're getting closer every time 👍

File: lib/cli/verifyLinks/getUrlsFromSitemap.ts [hunk 8]
```diff
@@ -1,15 +1,88 @@
...
+}
+
+interface SitemapContent {
+  urlset: { url: UrlSetEntry[] };
+}
+
+const invalidXMLStructureErrorMessage = 'Provided XML file has invalid structure.';
+
+const sitemapContentSchema = {
+  type: 'object' as const,
```
[the remark this answers is not in the document]
Comment (dotKuro) [reply lib/cli/verifyLinks/getUrlsFromSitemap.ts:17]: The problem with using `statement as type` is that it will skip the type checking for this type and asserts that it is that type. This will defeat the entire purpose of the library having types all together, because the type is supposed to validate whether or not we wrote a valid (valid in the sense of syntactically valid) JSON schema. Using this might be more annoying but is easier to debug for sure. 

Another solution would be to export constant tokens from the library it self.

in validate-value: 
```typescript
const stringValue = 'string'
const objectValue = 'object'
...



export {
...

  values:  {
    string: stringValue,
    object: objectValue
    ...
  }
}
```

in code:
```typescript
import { Value. values as v} from 'validate-value'

const schema = {
  type: v.object
  properties: { ... }
}

```

File: test/cli/getUrlsFromSitemapTests.ts (hunk 2 above)
Comment (dotKuro) [reply test/cli/getUrlsFromSitemapTests.ts:15]: Something like `2020-10-28T16:55:03.554Z` and is generated using `new Date().toISOString()`.

Should I add examples to the tests?

File: test/cli/getUrlsFromSitemapTests.ts (hunk 3 above)
Comment (dotKuro) [reply test/cli/getUrlsFromSitemapTests.ts:54]: I have no idea how to highlight it either. I personally think it is fine, since it is test data. I think for test data you usually realize rather quickly that stuff that looks fishy tests some weird edge cases.  But maybe that's only for me.

Comment (dotKuro) [reply test/cli/getUrlsFromSitemapTests.ts:54]: The only thing I could come up with is to replace the trailing newline with `\n` which is the same thing, but avoids that problem. This will be rather inconsistent with the other newlines in the middle of the data though.


Thoughts on that?

Comment (goloroden) [reply test/cli/getUrlsFromSitemapTests.ts:54]: After having thought about this, let's stick with it in the current way. It becomes clear when one reads the test description.

File: test/cli/getUrlsFromSitemapTests.ts (hunk 2 above)
Comment (goloroden) [reply test/cli/getUrlsFromSitemapTests.ts:15]: Yes, we could use an actual value here, to make things more clear. Theoretically, since it doesn't really matter which value is being used, we might actually generate one here on each run of the test.

File: lib/cli/verifyLinks/getUrlsFromSitemap.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (goloroden) [reply lib/cli/verifyLinks/getUrlsFromSitemap.ts:17]: Hmmm 🤔 

What would you prefer?

[the remark this answers is not in the document]
Comment (dotKuro) [reply lib/cli/verifyLinks/getUrlsFromSitemap.ts:17]: Both pretty much have the same disadvantage unfortunately: You can't just paste a valid JSON schema from somewhere else and have it work with out adjusting it. Ideally we would find a solution that would be able to do that, but for the time being I would prefer the exporting constant tokens variant over the `as const` variant. It is much better to read in my opinion.

[the remark this answers is not in the document]
Comment (goloroden) [reply lib/cli/verifyLinks/getUrlsFromSitemap.ts:17]: Then let's do it like this, using the `values` property. And I like the approach with the alias. 👍

**Comment by dotKuro** [comment]
Everything should be resolved now 👍

**Review by goloroden — APPROVED** [review]
